plugin/grpc/proxy.go

```
@@ -0,0 +1,146 @@
+package grpc
+
+import (
+	"context"
+	"crypto/tls"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"github.com/coredns/coredns/pb"
+	"github.com/coredns/coredns/plugin/pkg/log"
+	"github.com/coredns/coredns/plugin/pkg/up"
+	"github.com/miekg/dns"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/status"
+)
+
+// Proxy defines an upstream host.
+type Proxy struct {
+	fails uint32
+	addr  string
+
+	// connection
+	client   pb.DnsServiceClient
+	dialOpts []grpc.DialOption
+
+	// health checking
+	probe *up.Probe
+}
+
+// GrpcOpts defines an GrpcProxy options.
+type options struct {
+	TLSConfig *tls.Config
+}
+
+// newProxy returns a new proxy.
+func newProxy(addr string, opts *options) *Proxy {
+	p := &Proxy{
+		fails: 0,
+		addr:  addr,
+		probe: up.New(),
+	}
+
+	if opts.TLSConfig != nil {
+		p.dialOpts = append(p.dialOpts, grpc.WithTransportCredentials(credentials.NewTLS(opts.TLSConfig)))
+	} else {
+		p.dialOpts = append(p.dialOpts, grpc.WithInsecure())
+	}
+
+	conn, err := grpc.Dial(p.addr, p.dialOpts...)
+	if err != nil {
+		log.Warningf("Skipping gRPC host '%s' due to Dial error: %s\n", p.addr, err)
```

should probably just return the error instead of logging or continuing?